pyaiot/gateway/common/nodes_controller.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2017 IoT-Lab Team
+# Contributor(s) : see AUTHORS file
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice,
+# this list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+# this list of conditions and the following disclaimer in the documentation
+# and/or other materials provided with the distribution.
+#
+# 3. Neither the name of the copyright holder nor the names of its contributors
+# may be used to endorse or promote products derived from this software without
+# specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""Base class for nodes controllers."""
+
+import logging
+
+from abc import ABCMeta, abstractmethod
+from tornado import gen
+
+from pyaiot.common.messaging import Message as Msg
+
+logger = logging.getLogger("pyaiot.gw.common.nodes")
+
+
+class NodesControllerBase(metaclass=ABCMeta):
+    """Base nodes controller class."""
+
+    def __init__(self, on_message_cb):
+        self._on_message_cb = on_message_cb
+        self.nodes = {}
+
+    @abstractmethod
+    def send_data_to_node(self, data):
+        """Forward received message data to the destination node.
+
+        The message should be JSON and contain 'uid', 'path' and 'payload'
+        keys.
+
+        - 'uid' corresponds to the node uid (uuid)
+        - 'path' corresponds to the resource name on the node
+        - 'payload' corresponds to the new payload for the resource.
+        """
+
+    @gen.coroutine
+    def send_message_to_broker(self, msg):
+        """Send a message to the broker via the gateway."""
```

Comment:
A comment on what is supposed to be `msg` would be great.
I could guess it's quite the same as `data` in `send_data_to_node`, but not sure.
If it's the same structure, why not the same name ?

Reply:
> I could guess it's quite the same as data in send_data_to_node, but not sure.
If it's the same structure, why not the same name ?

Good point, I'll look into that and try to come with a better 'design': the specific node controllers should manipulate nodes with a defined interface (since they are used by this base class). For the moment they are only managed by a dict. I need to think of that.

Reply:
After looking at that, `msg` and `data` are 2 different things: the former is a serialized version of a message (serialized JSON) and the former is dict with required keys.

I'll update this PR accordingly.

Reply:
@astralien3000, comment addressed. Is it fine for you ?